### ISSUE TITLE
Any way to save an attached .msg file as an actual .msg file? 

### ISSUE THREAD

**Issue by felix0908** [TeamMsgExtractor/msg-extractor#219]
Hi!
I'm using your module to handle .msg files for an archival project, and it works pretty well for extracting attachments, even in nested .msg files. But for archival purposes I'd really like to save any .msg files AS .msg files, rather than turning them into plaintext representations - but I've been entirely unable to figure out how to do this! 

I know it's not really what this module sets out to do, but is there any way to accomplish this right now? 

I know it CAN be done, because when .msg files get nested enough this module sometimes misidentifies .msg files as simple "data" attachments, and then saves them as-is, which is exactly what I'd like it to do.

Any help is appreciated! And I apologize if this isn't the right place for this

**Comment by TheElementalOfDestruction** [comment]
Duplicate of https://github.com/TeamMsgExtractor/msg-extractor/issues/117 where it is talked about at least briefly. If you want the full explanation on the issue, continue reading below, but the TL;DR is that it is not simple and we are waiting for an update to olefile.

More in depth explanation is a go. Alright, so the way an MSG file works is that it is essentially a special type of archive, which we are using another module, OleFile, to open. The msg standard applies to the files and structures accessible from an interface with the archive, so all we have to do is open streams (files) and folders and go poking around to pull out the data. When it comes to embedded msg files, it's not as many people would expect. Rather than being a raw data stream that we could just drop into another file, it's actually a folder containing almost all of the streams of the original file, with one exception: named properties. The naked properties for all embedded msg files, all attachments, everything, is in the top level msg and purged from the embedded ones. The only thing that remains is the streams with the names data, but not the names of the properties. Extracting the msg file that was embedded would require some more advanced code to go through and untangle all the streams to create a new msg file from the embedded one. While that would suck to do, it's actually something I am quite willing to do at some point.

There is, however, a roadblock that makes the previous point moot. Regardless of my willingness to make the code that would go and untangle all the streams to reassemble the file, it would be useless without somewhere to put it all, which is where the new (well, actually it's a pretty old one) problem comes in: olefile. Olefile does not support the creation of new OleFile archives, something we would need to be able to create a new msg file in the first place. Without this basic building block, there is nowhere for us to go. We only have two options left: get olefile to the point where this is something that is supported or attempt to transition to a new, probably self written, module. As far as I'm aware, no other module for Python does what olefile can. And, despite having access to the technical documentation that would be the minimum requirement for implementing it, such a thing would take a lot more time, delaying every other feature while this one is worked on. 

As such, all I can suggest that you do if you want this feature is to go check on the progress from olefile and offer any help you can to getting them to the stage required for this feature.

**Comment by TheElementalOfDestruction** [comment]
Side note, extracting an attachment in its original I modified form would actually be part of the Gould for this module, so yes, this feature is something we want to implement but just can't.

**Comment by felix0908** [comment]
Thank you for the quick response! 

If this isn't something that's actually possible to do (which I'm perfectly willing to believe you about), then do you have any possible explanations for the "bugged" behavior where a nested .msg file IS saved, rather than handled as an embedded message? Because this seems very curious

**Comment by TheElementalOfDestruction** [comment]
My best guess is that it's likely not a bug, but rather a limit to the file. I believe in the cursory glance I took at the format that the paths have a limit to the total length, or the amount of times they can be nested. As such, if they were to try to next any further they may end up invalidating the file, and so have to stop at a certain point. Microsoft of course has a much better understanding of all the formats involved than could likely ever be gleaned from the documentation (the documentation for undeniable recipients blatantly does not match what I have seen outlook produce for it, so I honestly have no idea how to handle that) so when you try to next a file that's already nested, they likely were able to easily code a way for it to partially unnest and detangle itself. I can't stress this enough, but that is so much easier when you wrote the format than if you have some inconsistent and confusing documentation that also manages to frequently be plain omitting information that could be extremely useful.

TL;DR: probably the same reason that you can't have too long a file path on windows or something. Microsoft format means they can handle it easily.

**Comment by felix0908** [comment]
Ahhh, okay, so you think it might be a quirk in .msg files themselves - that's definitely possible.
And yes, I understand quite well - we run into weird, inconsistently-designed and proprietary formats fairly frequently, and figuring out how to handle them archivally is an eternal problem. One of the reasons it's always nice to find a module like this!

Thank you for your time, and your elaborate answers

**Comment by TheElementalOfDestruction** [comment]
Update: This is now completely possible after I wrote a new class specifically for the purpose of creating a new OLE file. Version 0.38.0 contains this ability.
